text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,335 @@
+
+- [Svelte: `<svelte:component this={expression}>`](https://svelte.dev/docs#template-syntax-svelte-component)
+- [Vue: `<component :is="...">`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
+- [Angular: viewContainerRef](https://angular.io/guide/dynamic-component-loader)
+- [Lit: static expressions](https://lit.dev/docs/templates/expressions/#static-expressions)
+- [Aura: $A.createComponent](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_cmp_async.htm)
+
+Frameworks that use JSX such as [React](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom) and [Stencil](https://stenciljs.com/docs/templating-jsx) can dynamically render components directly into the markup using expressions.
+
+The goal of managing dynamic components in LWC is to follow the same component lifecycle as all other LWC custom elements.  Having said that, taking a similar approach as Svelte and Vue seems the most natural and intuitive path forward.
+
+## Detailed design
+
+A new property, `lightningElementName` on `LightningElementConstructor` is introduced to store the name of the custom element.
+
+A special tag, `<lwc.dynamic>` along with a new template directive `lwc:is` are also introduced and serves as the anchor in the DOM where the dynamic component will be rendered.
+
+Broken down, there are three main parts to this design:
+
+1. Storing and retrieving a custom element's name.
+2. Defining an anchor in the DOM.
+3. Instantiating the dynamic custom element.
+
+#### Storing and retrieving a custom element's name
+
+In today's world, an LWC module's namespace and name are known at compile time by [examining the directory structure](https://rfcs.lwc.dev/rfcs/lwc/0101-module-unification) and through [module resolution](https://rfcs.lwc.dev/rfcs/lwc/0020-module-resolution).
+
+The default folder structure for an LWC module looks as follows:
+
+```
+namespace/
+└── name/
+    ├── *.css
+    ├── *.js
+    ├── *.html
+```
+
+Leveraging this, the namespace and name can be accessed at compile time and ultimately stored on the constructor.  A new property on `LightningElementConstructor` called `lightningElementName` will store the custom element name which will be of the form `namespace-name`.
+
+```javascript
+const { default: ctor } = await import("c/customConstructor");
+// The tag name can be directly accessed on the constructor
+ctor.lightningElementName
+```
+
+The value of `lightningElementName` should only be set by the LWC engine to prevent component authors from modifying it and producing unintended side effects.
+
+Alternatively, the custom element name can be stored in a `weakMap` and a public API can be exposed to retrieve the constructor's custom element name. However, because the custom element name can be used as a selector with `this.template.querySelector` placing it on the constructor seems more ergonomic.
```

Comment:
I would strongly suggest we go with the `WeakMap` approach, so that it's not exposed to component authors. `this.template.querySelector` should be unnecessary; component authors can use template refs instead. (Assuming we make template refs work with this proposal. 🙂)

Reply:
Updated to use a `Map` since that's what's [currently being used for templates](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/component.ts#L18).